### PR TITLE
Rchain 1117: make purse deposit fail on overflow

### PR DESCRIPTION
**Pull request by odeac**
## Overview
Make purse deposit fail on overflow

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-1117

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/rholang/MakeMint.rho [hunk 1]
```diff
@@ -47,18 +49,27 @@ in {
...
+              new ch in {
+                balance!("add", amount, *ch) |
+                for(@addSuccess <- ch) {
+                  if(addSuccess) {
+                    @src!(*decr, amount, *ch) |
+                    for(@decrSuccess <- ch) {
+                      if (decrSuccess) {
+                        successCh!(true)
+                      } else {
+                        balance!("sub", amount, *successCh)
```
Comment (KentShikama) [inline casper/src/main/rholang/MakeMint.rho]: So in case you fail, you subtract the amount that you added? Why not just never add the amount in the first place?

File: casper/src/test/rholang/FailingResultCollectorTest.rho [hunk 2]
```diff
@@ -1,8 +1,16 @@
...
-new assert(`rho:test:assert`), assertEquals(`rho:test:assertEquals`)  in {
+new assert(`rho:test:assertAck`),
+    testCompleted(`rho:test:testSuiteCompleted`),
+    ack
+in {
+  assert!("false should fail", false, "clue assertTrue", *ack) |
 
-  assert!("false should fail", false) |
+  for (_ <- ack) {
+    assert!("0 != 1 should fail", (0, "==", 1), "clue assertEquals", *ack) |
```
Comment (KentShikama) [inline casper/src/test/rholang/FailingResultCollectorTest.rho]:
```suggestion
    assert!("0 == 1 should fail", (0, "==", 1), "clue assertEquals", *ack) |
```

File: casper/src/main/rholang/MakeMint.rho [hunk 3]
```diff
@@ -47,18 +49,27 @@ in {
...
-            contract thisPurse(@"deposit", @amount, @src, success) = {
-              new result in {
-                @src!(*decr, amount, *result) | //call src decr function.
-                for(@decrSuccess <- result) {
-                  if (decrSuccess) {
-                    balance!("add", amount, *success) // add transferred amount to this purse
+
+            contract thisPurse(@"deposit", @amount, @src, successCh) = {
+              new ch in {
+                balance!("add", amount, *ch) |
```
Comment (dckc) [inline casper/src/main/rholang/MakeMint.rho]: NonNegativeNumber already checks for overflow. I don't see why this change is needed.

Comment (dckc) [reply casper/src/main/rholang/MakeMint.rho]: oops; now I see; never mind...

File: casper/src/main/rholang/MakeMint.rho [hunk 4]
```diff
@@ -47,18 +49,27 @@ in {
...
+                for(@addSuccess <- ch) {
+                  if(addSuccess) {
+                    @src!(*decr, amount, *ch) |
```
Comment (dckc) [inline casper/src/main/rholang/MakeMint.rho]: All possible failure modes have to be checked **before** the call to "decr".

See https://github.com/Agoric/PlaygroundVat/blob/master/examples/contract/makeMint.js#L42

File: casper/src/main/rholang/MakeMint.rho [hunk 5]
```diff
@@ -47,18 +49,27 @@ in {
...
               }
             } |
-            contract thisPurse(@"deposit", @amount, @src, success) = {
-              new result in {
-                @src!(*decr, amount, *result) | //call src decr function.
-                for(@decrSuccess <- result) {
-                  if (decrSuccess) {
-                    balance!("add", amount, *success) // add transferred amount to this purse
+
+            contract thisPurse(@"deposit", @amount, @src, successCh) = {
```
Comment (dckc) [inline casper/src/main/rholang/MakeMint.rho]: Is changing `success` to `successCh` worthwhile? Is there a style guide we're trying to follow? It made the diff hard for me to read.

Comment (odeac) [reply casper/src/main/rholang/MakeMint.rho]: I'm not sure if we have a style guide but we should. I wanted to make it clear that the parameter is the return channel

File: casper/src/main/rholang/MakeMint.rho (hunk 4 above)
Comment (odeac) [reply casper/src/main/rholang/MakeMint.rho]: what other failure modes are you thinking about?

File: casper/src/main/rholang/MakeMint.rho (hunk 1 above)
Comment (odeac) [reply casper/src/main/rholang/MakeMint.rho]: The failure comes from the balance, which is a NonNegativeNumber, if I was to check before add I would duplicate the code in NonNegativeNumber

File: casper/src/test/rholang/FailingResultCollectorTest.rho (hunk 2 above)
Comment (odeac) [reply casper/src/test/rholang/FailingResultCollectorTest.rho]: this was rebased on top of PR 2100 because I needed the tests. Please ignore the rest of the changes except MakeMint.rho and MakeMintTest.rho

File: casper/src/main/rholang/MakeMint.rho (hunk 4 above)
Comment (dckc) [reply casper/src/main/rholang/MakeMint.rho]: Not other failure modes but the order of checking.

I'm saying something similar to what Kent is saying: check before changing any balances. Don't change and then undo.

Comment (dckc) [reply casper/src/main/rholang/MakeMint.rho]: Also... we need to make sure there is no observable state in between increasing the destination balance and decreasing the source balance. Somehow we need to do something like:

```
  for (@destBal <- destBalCh; @srcBal <- srcBalCh) {
    // check that srcBal >= amount and that destBal + amount doesn't overflow
    if (okToCommit) {
      destBalCh!(destBal + amount) | srcBalCh!(srcBal - amount)
    } else {
      destBal!(destBal) | srcBal!(srcBal)
    }
  }
```

In fact, when we get peek, we shouldn't even consume the balances from their channels until the commit point.

Comment (odeac) [reply casper/src/main/rholang/MakeMint.rho]: >> we need to make sure there is no observable state in between 
This is a very good point!

File: casper/src/test/rholang/EitherTest.rho [hunk 6]
```diff
@@ -1,82 +1,81 @@
...
-  rl(`rho:registry:lookup`), EitherCh, TestSetCh,
+new
+  rl(`rho:registry:lookup`), EitherCh, RhoSpecCh,
+  stdlog(`rho:io:stdlog`),
+  double, divide, divide10By, divide8By,
+  /* TestSetCh,
   double, divide, divide100by, divide50by, divide10by,
-  divide8by, doubleMap, divide10byFlatMap, composeTest,
-  test1, test2, test3
+  divide8by, doubleMap, divide10byFlatMap, composeTest, */
```
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/rholang/EitherTest.rho]: this is due to the rebase on top of PR 2100. Will fix these after 2100 is merged

File: casper/src/main/rholang/MakeMint.rho (hunk 4 above)
Comment (dckc) [reply casper/src/main/rholang/MakeMint.rho]: probably should be its own JIRA ticket

Comment (odeac) [reply casper/src/main/rholang/MakeMint.rho]: Still processing...

Comment (odeac) [reply casper/src/main/rholang/MakeMint.rho]: I've created a new JIRA ticket for this: https://rchain.atlassian.net/browse/RCHAIN-2926
